File: ios/Shared/TextDocumentProxyProtocol.swift
import UIKit

/**
 * TextDocumentProxyProtocol - Abstraction over UITextDocumentProxy
 *
 * This protocol allows KeyboardEngine to work with different text input sources:
 * - Real keyboard extensions use UITextDocumentProxy (iOS system)
 * - Preview uses CustomTextDocumentProxy (React Native bridge)
 *
 * Mirrors the essential parts of UITextDocumentProxy interface.
 */
protocol TextDocumentProxyProtocol: AnyObject {
    /// Text before the cursor/insertion point
    var documentContextBeforeInput: String? { get }

    /// Text after the cursor/insertion point
    var documentContextAfterInput: String? { get }

    /// Whether there is text in the document or a text selection
    var hasText: Bool { get }

    /// Insert text at cursor position
    func insertText(_ text: String)

    /// Delete one user-visible character backward (a whole grapheme cluster,
    /// e.g. a Hebrew letter together with its nikkud)
    func deleteBackward()

    /// Delete one Unicode scalar backward — used to strip a single combining mark
    /// (e.g. replacing one nikkud vowel with another) without removing its base letter.
    func deleteScalarBackward()

    /// Adjust cursor position by character offset
    func adjustTextPosition(byCharacterOffset offset: Int)

    /// Keyboard type hint (for field type detection)
    var keyboardType: UIKeyboardType? { get }

    /// Return key type (Go, Search, Done, etc.)
    var returnKeyType: UIReturnKeyType? { get }

    /// Text content type (email, URL, etc.)
    @available(iOS 10.0, *)
    var textContentType: UITextContentType? { get }
}

extension TextDocumentProxyProtocol {
    /// Default: UIKit's deleteBackward already removes a single scalar when the
    /// preceding character is a combining mark, so the system proxy needs no override.
    /// Only CustomTextDocumentProxy, which edits a Swift String, must distinguish them.
    func deleteScalarBackward() {
        deleteBackward()
    }
}

File: android/app/src/main/java/org/issieshapiro/issieboard/shared/TextDocumentProxyProtocol.kt
package org.issieshapiro.issieboard.shared

/**
 * Port of ios/Shared/TextDocumentProxyProtocol.swift
 *
 * TextDocumentProxyProtocol - Abstraction over Android InputConnection
 *
 * This interface allows KeyboardEngine to work with different text input sources:
 * - Real keyboard services use InputConnection (Android system)
 * - Preview uses CustomTextDocumentProxy (React Native bridge)
 *
 * Mirrors the essential parts of InputConnection interface.
 */
interface TextDocumentProxyProtocol {
    /**
     * Text before the cursor/insertion point
     */
    val documentContextBeforeInput: String?

    /**
     * Text after the cursor/insertion point
     */
    val documentContextAfterInput: String?

    /**
     * Whether there is text in the document or a text selection
     */
    val hasText: Boolean

    /**
     * Insert text at cursor position
     */
    fun insertText(text: String)

    /**
     * Delete one character backward
     */
    fun deleteBackward()

    /**
     * Adjust cursor position by character offset
     */
    fun adjustTextPosition(offset: Int)

    /**
     * Keyboard type hint (for field type detection)
     * Maps to Android InputType
     */
    val keyboardType: Int?

    /**
     * Return key type (Go, Search, Done, etc.)
     * Maps to Android ImeOptions
     */
    val returnKeyType: Int?

    /**
     * Text content type (email, URL, etc.)
     * Maps to Android autofillHints
     */
    val textContentType: String?
}
